cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -852,56 +859,116 @@ public void addSchedule(ApplicationId applicationId, ScheduleSpecification sched
+                                                      "name or data trigger configuration in schedule update " +
+                                                      "details %s. These configuration will be ignored due to type " +
+                                                      "mismatch.", existingScheduleSpec.getSchedule().getName(),
+                                                    ScheduleType.TIME, scheduleUpdateDetail));
+      }
+      TimeSchedule oldTimeSchedule = (TimeSchedule) existingScheduleSpec.getSchedule();
+      String cron = scheduleUpdateDetail.getCronExpression() == null ? oldTimeSchedule.getCronEntry() :
+        scheduleUpdateDetail.getCronExpression();
+      schedule = new TimeSchedule(existingScheduleSpec.getSchedule().getName(), desc, cron, new RunConstraints(runs));
+
+    } else if (ScheduleType.fromSchedule(existingScheduleSpec.getSchedule()) == ScheduleType.STREAM) {
+      if (scheduleUpdateDetail.getCronExpression() != null) {
+        throw new BadRequestException(String.format("Schedule %s being updated is of type %s and found cron " +
+                                                      "expression configuration in schedule update details %s. " +
+                                                      "This configuration will be ignored due to type mismatch.",
+                                                    existingScheduleSpec.getSchedule().getName(), ScheduleType.STREAM,
+                                                    scheduleUpdateDetail));
+      }
+      StreamSizeSchedule oldTimeSchedule = (StreamSizeSchedule) existingScheduleSpec.getSchedule();
+      String sName = scheduleUpdateDetail.getStreamName() == null ? oldTimeSchedule.getStreamName() :
+        scheduleUpdateDetail.getStreamName();
+      int dTrigger = scheduleUpdateDetail.getDataTriggerMB() == null ? oldTimeSchedule.getDataTriggerMB() :
+        scheduleUpdateDetail.getDataTriggerMB();
+      schedule = new StreamSizeSchedule(existingScheduleSpec.getSchedule().getName(), desc, sName, dTrigger);
+    } else {
+      throw new BadRequestException(String.format("Invalid schedule type %s",
+                                                  ScheduleType.fromSchedule(existingScheduleSpec.getSchedule())));
+    }
+
+    Map<String, String> prop = scheduleUpdateDetail.getProperties() == null ? existingScheduleSpec.getProperties() :
+      getUpdatedProperties(existingScheduleSpec.getProperties(), scheduleUpdateDetail.getProperties());
+    return new ScheduleSpecification(schedule, existingScheduleSpec.getProgram(), prop);
+  }
+
+  /**
+   * Returns a map of {@link String}, {@link String} which has all the unique keys with their values from existing
+   * and updated. For the keys which are similar in existing and updated it retains those keys and values from updated
+   * are kept.
+   */
+  private Map<String, String> getUpdatedProperties(Map<String, String> existing, Map<String, String> updated) {
```

Comment:
It would be good to not merge the properties, and instead replace the properties. This will enable us to delete properties.